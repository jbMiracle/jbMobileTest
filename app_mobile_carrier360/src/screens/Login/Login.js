import React, { Component, Fragment } from 'react';
import { View, TouchableWithoutFeedback } from 'react-native';
import { connectStyle, Text } from 'native-base';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import styleDef from './Login.style';
import LoginBiometrics from './LoginBiometrics/LoginBiometrics';
import localizedString from '../../services/localization';
import { Auth as AuthActions } from '../../state/actions';
import LoginPassword from './LoginPassword/LoginPassword';

const loginTextOptions = { scope: 'Login' };
const loginString = key => localizedString(key, loginTextOptions);

const LoginMethods = {
  Password: 'password',
  Biometric: 'biometric',
  Guest: 'track_load_as_guest',
};

const mapStateToProps = (state) => {
  const { isSupported, supportedType, isSetup } = state.auth.biometric;
  return {
    isBiometricLoginSupported: isSupported,
    supportedBiometryType: supportedType,
    isBiometricLoginSetup: isSetup,
  };
};
const mapDispatchToProps = dispatch => ({
  authenticated: result => dispatch(AuthActions.userAuthenticated(result)),
  setupBiometricLogin: () => dispatch(AuthActions.setupBiometricLogin()),
});

class Login extends Component {
  state = {
    selectedMethod: LoginMethods.Password,
  };

  componentWillMount() {
    const { isBiometricLoginSetup } = this.props;
    this.setState({
      selectedMethod: isBiometricLoginSetup ? LoginMethods.Biometric : LoginMethods.Password,
    });
  }

  onMethodSelected = method => () => this.setState({ selectedMethod: method });

  renderMethodSelector(method, labelKey = method) {
    const isSelected = this.state.selectedMethod === method;
    const fontStyle = { fontWeight: isSelected ? 'bold' : 'normal' };

    return (
      <TouchableWithoutFeedback testID='login_loginButton' onPress={ this.onMethodSelected(method) } style={ styleDef.methodButton }>
        <View style={ styleDef.methodButtonView }>
          <Text style={ [styleDef.methodText, fontStyle] }>{loginString(labelKey)}</Text>
          {isSelected && <View style={ styleDef.triStyle } />}
        </View>
      </TouchableWithoutFeedback>
    );
  }

  renderMethodContent() {
    const { Password, Biometric, Guest } = LoginMethods;
    const { selectedMethod } = this.state;

    switch (selectedMethod) {
      case Password:
        return <LoginPassword { ...this.props } />;
      case Biometric:
        return <LoginBiometrics { ...this.props } />;
      case Guest:
        return <View style={ { flex: 1 } } />;
      default:
        return null;
    }
  }

  render() {
    const { Password, Biometric, Guest } = LoginMethods;
    const { isBiometricLoginSupported, supportedBiometryType } = this.props;

    return (
      <View style={ styleDef.container }>
        {this.renderMethodContent()}
        <View style={ styleDef.methodsContainer }>
          {this.renderMethodSelector(Password)}
          <View style={ styleDef.methodSeparator } />
          {isBiometricLoginSupported && (
            <Fragment>
              {this.renderMethodSelector(Biometric, supportedBiometryType)}
              <View style={ styleDef.methodSeparator } />
            </Fragment>
          )}
          {this.renderMethodSelector(Guest)}
        </View>
      </View>
    );
  }
}

Login.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func.isRequired,
  }).isRequired,
  isBiometricLoginSupported: PropTypes.bool,
  supportedBiometryType: PropTypes.string,
  isBiometricLoginSetup: PropTypes.bool,
};
Login.defaultProps = {
  isBiometricLoginSupported: false,
  supportedBiometryType: null,
  isBiometricLoginSetup: false,
};

export default connect(mapStateToProps, mapDispatchToProps)(connectStyle('C360.Login', styleDef)(Login));
