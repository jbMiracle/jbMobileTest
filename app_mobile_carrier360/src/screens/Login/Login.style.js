import { StyleSheet } from 'react-native';
import C360Colors from '../../themes/360Colors';

export default {
  container: {
    alignItems: 'center',
    backgroundColor: C360Colors.Blue500,
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'center',
  },
  methodsContainer: {
    alignItems: 'flex-end',
    borderBottomColor: C360Colors.Blue600,
    borderBottomWidth: 12,
    flexDirection: 'row',
  },
  methodButton: {
    alignItems: 'center',
    flex: 1,
    justifyContent: 'center',
  },
  methodButtonView: {
    alignItems: 'center',
    flex: 1,
    height: 60,
    justifyContent: 'center',
    paddingHorizontal: 20,
  },
  methodSeparator: {
    backgroundColor: C360Colors.White,
    height: 30,
    marginVertical: 15,
    width: StyleSheet.hairlineWidth,
  },
  methodText: {
    color: C360Colors.White,
    fontSize: 14,
    textAlign: 'center',
  },
  triStyle: {
    backgroundColor: 'transparent',
    borderBottomColor: C360Colors.Blue600,
    borderBottomWidth: 12,
    borderLeftColor: 'transparent',
    borderLeftWidth: 12,
    borderRightColor: 'transparent',
    borderRightWidth: 12,
    borderStyle: 'solid',
    bottom: 0,
    height: 0,
    position: 'absolute',
    width: 0,
  },
};
