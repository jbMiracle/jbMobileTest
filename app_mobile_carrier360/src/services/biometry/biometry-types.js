export default {
  TouchID: 'touch_id',
  FaceID: 'face_id',
};
