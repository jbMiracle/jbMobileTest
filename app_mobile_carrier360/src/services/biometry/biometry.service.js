import {
  Platform,
  // Alert,
} from 'react-native';
import TouchID from 'react-native-touch-id';
import BiometryTypes from './biometry-types';

export async function queryBiometricSupport() {
  try {
    // On Android: boolean
    // On iOS: string of touch/face
    const supportResult = await TouchID.isSupported();
    const supportedType = Platform.select({
      android: supportResult ? BiometryTypes.TouchID : null,
      ios: BiometryTypes[supportResult],
    });
    return {
      isSupported: !!supportResult,
      supportedType,
    };
  } catch (err) {
    return {
      isSupported: false,
      supportedType: null,
    };
  }
}

// const alertResult = result => new Promise((resolve) => {
//   Alert.alert(
//     'Biometric Authentication Result!',
//     result,
//     [
//       { text: 'OK', onPress: resolve },
//     ],
//     { cancelable: false }
//   );
// });

export async function authenticate(message = 'Pretty Please!') {
  try {
    const result = await TouchID.authenticate(message);
    // await alertResult(JSON.stringify(result));
    return result;
  } catch (err) {
    // User cancelling dialog brings us here.. now what?
    // On failed touch attempt:
    //   Android immediately throws failure
    //   iOS gives the user another chance before yielding a result/failure
    //     The user can try again or tap 'use password' which then throws back "LAErrorUserFallback"(iOS) / "Touch ID Error"(Android)
    // alertResult(JSON.stringify(err, null, 2));
    return false;
  }
}
