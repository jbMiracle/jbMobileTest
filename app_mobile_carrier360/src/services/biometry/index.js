import BiometryTypes from './biometry-types';

export { BiometryTypes };
export * from './biometry.service';
