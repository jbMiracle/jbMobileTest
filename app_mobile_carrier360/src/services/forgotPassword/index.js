import sendForgotPassword from './forgotPassword.service';

export default sendForgotPassword;
