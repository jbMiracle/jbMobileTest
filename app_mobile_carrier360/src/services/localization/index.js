import strings from './localization.service';

export default strings;
