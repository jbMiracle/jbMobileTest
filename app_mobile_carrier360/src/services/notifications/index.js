export * from './notifications.service';
export { ActionCategoryIds } from './actionCategories';
