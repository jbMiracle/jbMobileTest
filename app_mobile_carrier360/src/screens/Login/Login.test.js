import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Login from './Login';
import { BiometryTypes } from '../../services/biometry';

it('can render an empty component', () => {
  const renderer = new ShallowRenderer();
  const state = {
    auth: {
      biometric: {
        isSupported: true,
        supportedType: BiometryTypes.FaceID,
        isSetup: false,
      },
    },
  };
  const props = {
    store: {
      getState: () => state,
      subscribe: () => {},
      dispatch: () => {},
    },
    isBiometricLoginSupported: true,
    supportedBiometryType: BiometryTypes.FaceID,
    isBiometricLoginSetup: false,
  };

  renderer.render(<Login { ...props } />);

  expect(renderer.getRenderOutput()).toMatchSnapshot();
});
