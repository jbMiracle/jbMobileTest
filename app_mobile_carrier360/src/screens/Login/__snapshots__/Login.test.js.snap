// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can render an empty component 1`] = `
<Styled(Login)
  authenticated={[Function]}
  isBiometricLoginSetup={false}
  isBiometricLoginSupported={true}
  setupBiometricLogin={[Function]}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "notify": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": null,
    }
  }
  supportedBiometryType="face_id"
  virtual={undefined}
/>
`;
